File: Kanban.cs
// Kanban.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

public enum Status
{
    [JsonPropertyName("To Do")]
    Todo,
    [JsonPropertyName("In Progress")]
    InProgress,
    [JsonPropertyName("Done")]
    Done
}

public enum Priority
{
    [JsonPropertyName("Низкий")]
    Low,
    [JsonPropertyName("Средний")]
    Medium,
    [JsonPropertyName("Высокий")]
    High
}

public record Task(
    int Id,
    string Title,
    string Description,
    Status Status,
    Priority Priority,
    string CreatedAt
);

public class KanbanBoard
{
    private List<Task> tasks = new();
    private int nextId = 1;

    public IReadOnlyList<Task> Tasks => tasks.AsReadOnly();

    public Task AddTask(string title, string desc, Priority priority)
    {
        var task = new Task(
            nextId,
            title,
            desc,
            Status.Todo,
            priority,
            DateTime.Now.ToString("o")
        );
        tasks.Add(task);
        nextId++;
        return task;
    }

    public Task? FindTask(int id) => tasks.FirstOrDefault(t => t.Id == id);

    public bool MoveTask(int id, Status newStatus)
    {
        var task = FindTask(id);
        if (task == null) return false;
        tasks.Remove(task);
        var updated = task with { Status = newStatus };
        tasks.Add(updated);
        return true;
    }

    public bool DeleteTask(int id) => tasks.RemoveAll(t => t.Id == id) > 0;

    public void SaveToFile(string filename)
    {
        var options = new JsonSerializerOptions { WriteIndented = true };
        string json = JsonSerializer.Serialize(tasks, options);
        File.WriteAllText(filename, json);
    }

    public void LoadFromFile(string filename)
    {
        if (!File.Exists(filename)) return;
        string json = File.ReadAllText(filename);
        var loaded = JsonSerializer.Deserialize<List<Task>>(json);
        if (loaded != null)
        {
            tasks = loaded;
            nextId = tasks.Any() ? tasks.Max(t => t.Id) + 1 : 1;
        }
    }
}

public static class Program
{
    private static string ReadString(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private static int ReadInt(string prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            if (int.TryParse(Console.ReadLine(), out int result))
                return result;
            Console.WriteLine("Введите число.");
        }
    }

    private static void PrintTask(Task task)
    {
        string emoji = task.Status switch
        {
            Status.Todo => "📌",
            Status.InProgress => "🔄",
            Status.Done => "✅",
            _ => "❓"
        };
        Console.WriteLine($"{emoji} #{task.Id} - {task.Title} ({task.Status})");
        Console.WriteLine($"   Описание: {task.Description}");
        Console.WriteLine($"   Приоритет: {task.Priority}, Создана: {task.CreatedAt}");
    }

    public static void Main()
    {
        var board = new KanbanBoard();
        try { board.LoadFromFile("kanban_data.json"); }
        catch { Console.WriteLine("Не удалось загрузить данные."); }

        while (true)
        {
            Console.WriteLine("\n===== КАНБАН-ДОСКА (C#) =====");
            Console.WriteLine("1. Показать все задачи");
            Console.WriteLine("2. Добавить задачу");
            Console.WriteLine("3. Переместить задачу (изменить статус)");
            Console.WriteLine("4. Удалить задачу");
            Console.WriteLine("5. Сохранить в файл");
            Console.WriteLine("6. Загрузить из файла");
            Console.WriteLine("0. Выход");
            string choice = ReadString("Выберите действие: ");

            switch (choice)
            {
                case "0": return;
                case "1":
                    if (!board.Tasks.Any()) Console.WriteLine("Задач нет.");
                    else foreach (var t in board.Tasks) PrintTask(t);
                    break;
                case "2":
                    string title = ReadString("Введите название: ");
                    if (string.IsNullOrWhiteSpace(title))
                    {
                        Console.WriteLine("Название не может быть пустым.");
                        continue;
                    }
                    string desc = ReadString("Введите описание: ");
                    Console.WriteLine("Выберите приоритет: 1-Низкий, 2-Средний, 3-Высокий");
                    string prio = ReadString("> ");
                    Priority priority = prio switch
                    {
                        "1" => Priority.Low,
                        "3" => Priority.High,
                        _ => Priority.Medium
                    };
                    var task = board.AddTask(title, desc, priority);
                    Console.WriteLine($"Задача добавлена с ID {task.Id}");
                    break;
                case "3":
                    int id = ReadInt("Введите ID задачи: ");
                    var found = board.FindTask(id);
                    if (found == null)
                    {
                        Console.WriteLine("Задача не найдена.");
                        continue;
                    }
                    Console.WriteLine($"Текущий статус: {found.Status}");
                    Console.WriteLine("Выберите новый статус: 1-To Do, 2-In Progress, 3-Done");
                    string newChoice = ReadString("> ");
                    Status newStatus = newChoice switch
                    {
                        "1" => Status.Todo,
                        "2" => Status.InProgress,
                        "3" => Status.Done,
                        _ => Status.Todo // fallback
                    };
                    if (board.MoveTask(id, newStatus))
                        Console.WriteLine("Статус обновлён.");
                    else
                        Console.WriteLine("Ошибка.");
                    break;
                case "4":
                    int delId = ReadInt("Введите ID задачи для удаления: ");
                    if (board.DeleteTask(delId)) Console.WriteLine("Задача удалена.");
                    else Console.WriteLine("Задача не найдена.");
                    break;
                case "5":
                    try { board.SaveToFile("kanban_data.json"); Console.WriteLine("Сохранено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "6":
                    try { board.LoadFromFile("kanban_data.json"); Console.WriteLine("Загружено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                default: Console.WriteLine("Неизвестная команда."); break;
            }
        }
    }
}

File: kanban.js
// kanban.js
const fs = require('fs').promises;
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const question = (prompt) => new Promise(resolve => rl.question(prompt, resolve));

class Task {
    constructor(id, title, description, status = 'To Do', priority = 'Средний', createdAt = new Date().toISOString()) {
        this.id = id;
        this.title = title;
        this.description = description;
        this.status = status;
        this.priority = priority;
        this.createdAt = createdAt;
    }
}

class KanbanBoard {
    constructor() {
        this.tasks = [];
        this.nextId = 1;
    }

    addTask(title, description, priority) {
        const task = new Task(this.nextId, title, description, 'To Do', priority);
        this.tasks.push(task);
        this.nextId++;
        return task;
    }

    findTask(id) {
        return this.tasks.find(t => t.id === id);
    }

    moveTask(id, newStatus) {
        const task = this.findTask(id);
        if (!task) return false;
        task.status = newStatus;
        return true;
    }

    deleteTask(id) {
        const index = this.tasks.findIndex(t => t.id === id);
        if (index === -1) return false;
        this.tasks.splice(index, 1);
        return true;
    }

    async saveToFile(filename = 'kanban_data.json') {
        await fs.writeFile(filename, JSON.stringify(this.tasks, null, 2));
    }

    async loadFromFile(filename = 'kanban_data.json') {
        try {
            const data = await fs.readFile(filename, 'utf8');
            const loaded = JSON.parse(data);
            this.tasks = loaded.map(t => Object.assign(new Task(0), t));
            this.nextId = this.tasks.reduce((max, t) => Math.max(max, t.id), 0) + 1;
        } catch (err) {
            if (err.code !== 'ENOENT') throw err;
        }
    }
}

function printTask(task) {
    const emoji = { 'To Do': '📌', 'In Progress': '🔄', 'Done': '✅' };
    console.log(`${emoji[task.status] || '❓'} #${task.id} - ${task.title} (${task.status})`);
    console.log(`   Описание: ${task.description}`);
    console.log(`   Приоритет: ${task.priority}, Создана: ${task.createdAt}`);
}

async function main() {
    const board = new KanbanBoard();
    await board.loadFromFile();

    while (true) {
        console.log('\n===== КАНБАН-ДОСКА (JavaScript) =====');
        console.log('1. Показать все задачи');
        console.log('2. Добавить задачу');
        console.log('3. Переместить задачу (изменить статус)');
        console.log('4. Удалить задачу');
        console.log('5. Сохранить в файл');
        console.log('6. Загрузить из файла');
        console.log('0. Выход');
        const choice = await question('Выберите действие: ');

        if (choice === '0') break;

        switch (choice) {
            case '1':
                if (board.tasks.length === 0) {
                    console.log('Задач нет.');
                } else {
                    board.tasks.forEach(printTask);
                }
                break;
            case '2': {
                const title = await question('Введите название: ');
                if (!title.trim()) {
                    console.log('Название не может быть пустым.');
                    continue;
                }
                const desc = await question('Введите описание: ');
                console.log('Выберите приоритет: 1-Низкий, 2-Средний, 3-Высокий');
                const prio = await question('> ');
                let priority;
                if (prio === '1') priority = 'Низкий';
                else if (prio === '3') priority = 'Высокий';
                else priority = 'Средний';
                const task = board.addTask(title, desc, priority);
                console.log(`Задача добавлена с ID ${task.id}`);
                break;
            }
            case '3': {
                const id = parseInt(await question('Введите ID задачи: '));
                const task = board.findTask(id);
                if (!task) {
                    console.log('Задача не найдена.');
                    continue;
                }
                console.log(`Текущий статус: ${task.status}`);
                console.log('Выберите новый статус: 1-To Do, 2-In Progress, 3-Done');
                const newChoice = await question('> ');
                let newStatus;
                if (newChoice === '1') newStatus = 'To Do';
                else if (newChoice === '2') newStatus = 'In Progress';
                else if (newChoice === '3') newStatus = 'Done';
                else {
                    console.log('Неверный выбор.');
                    continue;
                }
                if (board.moveTask(id, newStatus)) {
                    console.log('Статус обновлён.');
                } else {
                    console.log('Ошибка.');
                }
                break;
            }
            case '4': {
                const id = parseInt(await question('Введите ID задачи для удаления: '));
                if (board.deleteTask(id)) {
                    console.log('Задача удалена.');
                } else {
                    console.log('Задача не найдена.');
                }
                break;
            }
            case '5':
                try {
                    await board.saveToFile();
                    console.log('Сохранено.');
                } catch (err) {
                    console.log('Ошибка сохранения:', err.message);
                }
                break;
            case '6':
                try {
                    await board.loadFromFile();
                    console.log('Загружено.');
                } catch (err) {
                    console.log('Ошибка загрузки:', err.message);
                }
                break;
            default:
                console.log('Неизвестная команда.');
        }
    }
    rl.close();
}

main().catch(console.error);
